### PR TITLE
added tutorial section to use OMPL as a pre-processor for CHOMP

### PR DESCRIPTION
**Pull request by raghavendersahdev**
### Addition of tutorial for running OMPL as a pre-processor planner for CHOMP

This PR is a work done by Raghavender as a part of 2018 Google summer of code. As a part of this PR, CHOMP planning adapter can be used to optimize paths obtained from OMPL. 

### Checklist
- [ ] addition of tutorials for running OMPL to produce an initial guess for CHOMP
- [ ] corresponding PR for the relevant code is available here [PR#1012](https://github.com/ros-planning/moveit/pull/1012)

### CODE REVIEW

File: doc/chomp_planner/chomp_planner_tutorial.rst [hunk 1]
```diff
@@ -127,4 +127,36 @@ OMPL is a open source library for sampling based / randomized motion planning al
...
 
-For scenes containing obstacles, CHOMP often generates paths which do not prefer smooth trajectories by addition of some noise (*ridge_factor*) in the cost function for the dynamical quantities of the robot (like acceleration, velocity). CHOMP is able to avoid obstacle in most cases but also fails in some if it gets stuck in the local minima and might report a solution not found due to a naive initial guess for the trajectory. OMPL on the other hand generates collision free smooth paths in the presence of obstacles too.
\ No newline at end of file
+For scenes containing obstacles, CHOMP often generates paths which do not prefer smooth trajectories by addition of some noise (*ridge_factor*) in the cost function for the dynamical quantities of the robot (like acceleration, velocity). CHOMP is able to avoid obstacle in most cases but also fails in some if it gets stuck in the local minima and might report a solution not found due to a naive initial guess for the trajectory. OMPL on the other hand generates collision free smooth paths in the presence of obstacles too.
```
[the remark this answers is not in the document]
Comment (raghavendersahdev) [reply doc/chomp_planner/chomp_planner_tutorial.rst]: done, thanks Mike.

File: doc/chomp_planner/chomp_planner_tutorial.rst [hunk 2]
```diff
@@ -127,4 +127,36 @@ OMPL is a open source library for sampling based / randomized motion planning al
...
+
+Using OMPL as a pre-processor for CHOMP
+---------------------------------------
+Here, it is demonstrated that CHOMP can also be used as a post processing optimization technique for plans obtained by other planners like OMPL. This section has the instructions for performing this behaviour. The intuition behind this is that OMPL acts as the initial planner to produce an initial guess for CHOMP. CHOMP then takes this initial guess and further optimizes the trajectory. This somewhat gaurantees optimized trajectories.
```
[the remark this answers is not in the document]
Comment (raghavendersahdev) [reply doc/chomp_planner/chomp_planner_tutorial.rst]: done

File: doc/chomp_planner/chomp_planner_tutorial.rst [hunk 3]
```diff
@@ -127,4 +127,36 @@ OMPL is a open source library for sampling based / randomized motion planning al
...
+To achieve this, follow the steps.
+
+#. Open the ``ompl_planning_pipeline.launch`` file in the ``<robot_moveit_config>/launch`` folder of your robot. For the panda robot it is `this <https://github.com/ros-planning/panda_moveit_config/blob/master/launch/ompl_planning_pipeline.launch.xml>`_ file. Edit this launch file, find the lines where ``<arg name="planning_adapters">`` is mentioned and change it to: ::
+
+    <arg name="planning_adapters" value="default_planner_request_adapters/AddTimeParameterization
+                   default_planner_request_adapters/FixWorkspaceBounds
+                   default_planner_request_adapters/FixStartStateBounds
+                   default_planner_request_adapters/FixStartStateCollision
+                   default_planner_request_adapters/FixStartStatePathConstraints
+
```
[the remark this answers is not in the document]
Comment (raghavendersahdev) [reply doc/chomp_planner/chomp_planner_tutorial.rst]: done

File: doc/chomp_planner/chomp_planner_tutorial.rst [hunk 4]
```diff
@@ -127,4 +127,36 @@ OMPL is a open source library for sampling based / randomized motion planning al
...
+                   default_planner_request_adapters/CHOMPOptimizerAdapter" />
+
+#. The values of the planning_adapters is the order in which the mentioned adapters are called / invoked. Order here matters. Inside the CHOMP adapter, a `call <https://github.com/raghavendersahdev/moveit/blob/CHOMP_planning_request_adapter/moveit_ros/planning/planning_request_adapter_plugins/src/chomp_optimizer_adapter.cpp#L183>`_ to OMPL is made before invoking the CHOMP optimization sover, so CHOMP takes the initial path computed by OMPL as the starting point to further optimize it.
```
[the remark this answers is not in the document]
Comment (raghavendersahdev) [reply doc/chomp_planner/chomp_planner_tutorial.rst]: done

File: doc/chomp_planner/chomp_planner_tutorial.rst [hunk 5]
```diff
@@ -127,4 +127,36 @@ OMPL is a open source library for sampling based / randomized motion planning al
...
+                   default_planner_request_adapters/FixWorkspaceBounds
+                   default_planner_request_adapters/FixStartStateBounds
+                   default_planner_request_adapters/FixStartStateCollision
+                   default_planner_request_adapters/FixStartStatePathConstraints
+
+                   default_planner_request_adapters/CHOMPOptimizerAdapter" />
+
+#. The values of the planning_adapters is the order in which the mentioned adapters are called / invoked. Order here matters. Inside the CHOMP adapter, a `call <https://github.com/raghavendersahdev/moveit/blob/CHOMP_planning_request_adapter/moveit_ros/planning/planning_request_adapter_plugins/src/chomp_optimizer_adapter.cpp#L183>`_ to OMPL is made before invoking the CHOMP optimization sover, so CHOMP takes the initial path computed by OMPL as the starting point to further optimize it. 
+
+#. Find the line where ``<rosparam command="load" file="$(find panda_moveit_config)/config/ompl_planning.yaml"/>`` is mentioned and replace it with: ::
```
[the remark this answers is not in the document]
Comment (raghavendersahdev) [reply doc/chomp_planner/chomp_planner_tutorial.rst]: done, yes thats better, thanks

File: doc/chomp_planner/chomp_planner_tutorial.rst [hunk 6]
```diff
@@ -127,4 +127,36 @@ OMPL is a open source library for sampling based / randomized motion planning al
...
+
+    <rosparam command="load" file="$(find panda_moveit_config)/config/ompl_planning.yaml"/>
```
[the remark this answers is not in the document]
Comment (raghavendersahdev) [reply doc/chomp_planner/chomp_planner_tutorial.rst]: done

File: doc/chomp_planner/chomp_planner_tutorial.rst [hunk 7]
```diff
@@ -127,4 +127,36 @@ OMPL is a open source library for sampling based / randomized motion planning al
...
+    <rosparam command="load" file="$(find panda_moveit_config)/config/chomp_planning.yaml"/>
+
+#. Doing these additions basically adds a CHOMP Optimization adapter and loads up the corresponding CHOMP planner's parameters, if not included default CHOMP parameter values would be used for the CHOMP Optimization adapter. To do this with your own robot replace ``panda_moveit_config`` to ``robot_moveit_config`` of your robot.
```
[the remark this answers is not in the document]
Comment (raghavendersahdev) [reply doc/chomp_planner/chomp_planner_tutorial.rst]: done

File: doc/chomp_planner/chomp_planner_tutorial.rst [hunk 8]
```diff
@@ -127,4 +127,36 @@ OMPL is a open source library for sampling based / randomized motion planning al
...
+
+#. Also ensure that in the ``move_group.launch`` file of ``<robot_moveit_config>/launch`` folder for your robot, the defaut planner is ``ompl``. If not change it to ``ompl``.
```
[the remark this answers is not in the document]
Comment (raghavendersahdev) [reply doc/chomp_planner/chomp_planner_tutorial.rst]: done

[the remark this answers is not in the document]
Comment (raghavendersahdev) [reply doc/chomp_planner/chomp_planner_tutorial.rst]: done, thanks, sure I will keep this in mind.

**Comment by davetcoleman** [comment]
@mlautman @raghavendersahdev can someone look into why travis is failing? i realize its unrelated to this PR

**Comment by mlautman** [comment]
@davetcoleman on it

**Comment by raghavendersahdev** [comment]
resolved.

**Comment by raghavendersahdev** [comment]
currently Travis is failing because I am linking [this line](https://github.com/raghavendersahdev/moveit/blob/CHOMP_planning_request_adapter/moveit_ros/planning/planning_request_adapter_plugins/src/chomp_optimizer_adapter.cpp#L183) [here](https://github.com/ros-planning/moveit_tutorials/pull/204/files#diff-91bdb56bddeacee840b00abd169ee8d0R146) which does not exist right on the `ros-planning/moveit kinetic` branch, but after the merge of this [PR #1012]([url](https://github.com/ros-planning/moveit/pull/1012)), it should be fine.

File: doc/chomp_planner/chomp_planner_tutorial.rst [hunk 9]
```diff
@@ -111,20 +111,54 @@ CHOMP has some optimization parameters associated with it. These can be modified
...
\ No newline at end of file
+For scenes containing obstacles, CHOMP often generates paths which do not prefer smooth trajectories by addition of some noise (*ridge_factor*) in the cost function for the dynamical quantities of the robot (like acceleration, velocity). CHOMP is able to avoid obstacle in most cases but it can fail if it gets stuck in the local minima due to a bad initial guess for the trajectory. OMPL can be used to generate collision-free seed trajectories for CHOMP to mitigate this issue.
+
+Using OMPL as a pre-processor for CHOMP
+---------------------------------------
+Here, it is demonstrated that CHOMP can also be used as a post-processing optimization technique for plans obtained by other planning algorithms. The intuition behind this is that some randomized planning algorithm produces an initial guess for CHOMP. CHOMP then takes this initial guess and further optimizes the trajectory. 
+To achieve this, follow the steps.
```
Comment (davetcoleman) [inline doc/chomp_planner/chomp_planner_tutorial.rst]: use ``:`` at end of sentence

File: doc/chomp_planner/chomp_planner_tutorial.rst [hunk 10]
```diff
@@ -111,20 +111,54 @@ CHOMP has some optimization parameters associated with it. These can be modified
...
+
+#. Open the ``ompl_planning_pipeline.launch`` file in the ``<robot_moveit_config>/launch`` folder of your robot. For the panda robot it is `this <https://github.com/ros-planning/panda_moveit_config/blob/master/launch/ompl_planning_pipeline.launch.xml>`_ file. Edit this launch file, find the lines where ``<arg name="planning_adapters">`` is mentioned and change it to: ::
```
Comment (davetcoleman) [inline doc/chomp_planner/chomp_planner_tutorial.rst]: Panda

File: doc/chomp_planner/chomp_planner_tutorial.rst [hunk 11]
```diff
@@ -111,20 +111,54 @@ CHOMP has some optimization parameters associated with it. These can be modified
...
+Choosing parameters for CHOMP requires some sort of intuition based on the environment we are working in. One can have the default parameters for CHOMP and this works well in environments without  obstacles. However in cases where the scene is populated with obstacles, we need to vary some parameters to ensure that CHOMP is not stuck in local minima, or quickly finds optimal solutions, preferring trajectories which avoids obstacles. Some parameters like increasing the *ridge_factor* to say 0.001 makes CHOMP avoids obstacles by not preferring smooth trajectories, so there is a trade-off  between smoothness and CHOMP's ability to avoid obstacles. Choosing the correct number of *max_iterations*, *learning_rate* is important based on the environment we are working in. Not choosing the appropriate CHOMP parameters might lead to CHOMP reporting not finding a collision-free path. *collision_clearance*, *collision_threshold* parameters are useful in specifying the minimum distance to be kept from obstacles to avoid collisions.
```
Comment (mlautman) [inline doc/chomp_planner/chomp_planner_tutorial.rst]: fix double space "without  obstacles"

File: doc/chomp_planner/chomp_planner_tutorial.rst [hunk 12]
```diff
@@ -111,20 +111,54 @@ CHOMP has some optimization parameters associated with it. These can be modified
...
+OMPL is a open source library for sampling based / randomized motion planning algorithms. Sampling based algorithms are probabilistically complete: a solution would be eventually found if one exists, however non-existence of a solution cannot be reported. These algorithms are efficient and usually find a solution quickly. OMPL does not contain any code related to collision checking or visualization as the designers of OMPL did not want to tie it to a any particular collision checker or visualization front end. The library is designed so it can be easily integrated into systems that provide the additional components. MoveIt integrates directly with  OMPL and uses the motion planners from OMP as its default set of planners. The planners in OMPL are abstract; i.e. OMPL has no concept of a robot. Instead, MoveIt! configures OMPL and provides the back-end for OMPL to work with problems in Robotics.
```
Comment (mlautman) [inline doc/chomp_planner/chomp_planner_tutorial.rst]: OMPL "from OMP as"

Comment (raghavendersahdev) [reply doc/chomp_planner/chomp_planner_tutorial.rst]: done

File: doc/chomp_planner/chomp_planner_tutorial.rst (hunk 11 above)
Comment (raghavendersahdev) [reply doc/chomp_planner/chomp_planner_tutorial.rst]: done

File: doc/chomp_planner/chomp_planner_tutorial.rst (hunk 10 above)
Comment (raghavendersahdev) [reply doc/chomp_planner/chomp_planner_tutorial.rst]: done

File: doc/chomp_planner/chomp_planner_tutorial.rst (hunk 9 above)
Comment (raghavendersahdev) [reply doc/chomp_planner/chomp_planner_tutorial.rst]: done

**Comment by raghavendersahdev** [comment]
Hi @davetcoleman , I closed this PR because after the merge of PR #220 this PR was rendered redundant as #220 included this PR's changes in it.

**Pull request closed without merging**